zerver/management/commands/realm_emoji.py

```diff
@@ -56,11 +57,12 @@ def handle(self, *args: Any, **options: str) -> None:
             if img_path is None:
                 self.print_help("./manage.py", "realm_emoji")
                 sys.exit(1)
-            emoji_filename = get_emoji_file_name(os.path.basename(img_path), name)
-            realm_emoji = check_add_realm_emoji(realm, name, emoji_filename)
+            realm_emoji = check_add_realm_emoji(realm, name)
+            emoji_filename = get_emoji_file_name(os.path.basename(img_path), str(realm_emoji.id))
             with open(img_path, 'rb') as fp:
                 upload_backend.upload_emoji_image(
                     fp, emoji_filename, realm.get_admin_users().first())
+                add_realm_emoji_filename_and_notify(realm_emoji, emoji_filename)
```

Comment:
@timabbott 

>Also, shouldn't this use add_realm_emoji_and_notify to notify clients?

Fixed this!

>Structurally, I think it'd be better for check_add_realm_emoiji to handle the filename gymnastics to avoid this sort >of bug; is there a reason you split that function?

The reason I split that function this way because I thought that not doing so would lead to a race condition in which the client would be notified of the new realm emoji even before the image file gets uploaded which will lead to a 404 error while fetching image for that emoji. Does that make sense to you or is there some other better way?

Reply:
I think we can prevent that, as long as we do the proper gymnastics in `check_add_realm_emoji`.  My view is that function should accept the file object as one of its arguments and do the whole `upload_backend` dance that you currently have in `zerver/views/realm_emoji.py`, providing the interface to its callers that either the thing is uploaded and the notification is sent, or nothing happens.

The reason that is a better design is that then if code in tests or a management command or whatever needs to add an emoji, it can just call `check_add_realm_emoji`, without needing to copy code from the view file.

Reply:
Agree, going to change it!